### PR TITLE
Add API to set arbitrary WebContents as devtools

### LINKED ISSUE #3177: can we have dockable developer tool bar for webview in atom - Browser-Plus
I found it will be nice to have a dockable developer toolbar when using the webview. Here is the issue raised in browser-plus(atom package)

https://github.com/skandasoft/browser-plus/issues/40

### PR DESCRIPTION
**Pull request by zcbenz**
Close #3177.

### `contents.setDevToolsWebContents(devToolsWebContents)`

* `devToolsWebContents` WebContents

Uses the `devToolsWebContents` as the target `WebContents` to show devtools.

The `devToolsWebContents` must not have done any navigation, and it should not
be used for other purposes after the call.

By default Electron manages the devtools by creating an internal `WebContents`
with native view, which developers have very limited control of. With the
`setDevToolsWebContents` method, developers can use any `WebContents` to show
the devtools in it, including `BrowserWindow`, `BrowserView` and `<webview>`
tag.

An example of showing devtools in a `<webview>` tag:

```html
<html>
<head>
  <style type="text/css">
    * { margin: 0; }
    #browser { height: 70%; }
    #devtools { height: 30%; }
  </style>
</head>
<body>
  <webview id="browser" src="https://github.com"></webview>
  <webview id="devtools"></webview>
  <script>
    const browserView = document.getElementById('browser')
    const devtoolsView = document.getElementById('devtools')
    browserView.addEventListener('dom-ready', () => {
      const browser = browserView.getWebContents()
      browser.setDevToolsWebContents(devtoolsView.getWebContents())
      browser.openDevTools()
    })
  </script>
</body>
</html>
```

An example of showing devtools in a `BrowserWindow`:

```js
const {app, BrowserWindow} = require('electron')

let win = null
let devtools = null

app.once('ready', () => {
  win = new BrowserWindow()
  win.loadURL('https://github.com')
  win.show()
  devtools = new BrowserWindow({show: false})
  devtools.show()
  win.webContents.setDevToolsWebContents(devtools.webContents)
  win.webContents.openDevTools({mode: 'detach'})
})
```

screen shot 2017-12-01 at 11 59 10

### CODE REVIEW

**Comment by deepak1556** [comment]
Should we trigger the destruction of `external_web_contents` when `managed_web_contents` is destroyed, so that it follows the default devtools behavior which closes when the inspected window is destroyed.

I guess this also enabled inspecting devtools window https://github.com/electron/electron/issues/2835 , SGTM 👍

**Comment by zcbenz** [comment]
> Should we trigger the destruction of external_web_contents when managed_web_contents is destroyed, so that it follows the default devtools behavior which closes when the inspected window is destroyed.

The `external_web_contents` is usually managed by some wrapper classes like `BrowserWindow` or `<webview>`, having `external_web_contents` destroyed would leave the wrappers in a corrupted state. I think it is fine to have developers handle the lifetime of devtools, I'll make it clear in the docs.

File: docs/api/web-contents.md [hunk 1]
```diff
@@ -1089,6 +1089,9 @@ with native view, which developers have very limited control of. With the
 the devtools in it, including `BrowserWindow`, `BrowserView` and `<webview>`
 tag.
 
+Note that closing the devtools does not destory the `devToolsWebContents`, it
+is caller's responsibility to close `devToolsWebContents`.
```
Comment (jkleinsc) [inline docs/api/web-contents.md]: @zcbenz shouldn't this be:
Note that closing the devtools does not destroy the `devToolsWebContents`, it 
is caller's responsibility to destroy `devToolsWebContents`.

**Comment by LeoMartinDev** [comment]
When is it available on npm ?

**Comment by zcbenz** [comment]
@LeoMartinDev It will be in next minor/major release, which can take some time since the master branch is not very stable yet.

**Comment by jacklovepdf** [comment]
@zcbenz @MarshallOfSound 
I develop a tool with electron. There are two webview components in the app. One shows the html page, another one shows the chrome devtools which debug the page. The current solution is:
```
const browserContents = browserView.getWebContents()
const devtoolsContents = devtoolsView.getWebContents()

browserContents.setDevToolsWebContents(devtoolsContents)
browserContents.debugger.attach();

browserContents.openDevTools({
     detach: true
})

browserContents.debugger.sendCommand(    // not work
          'Emulation.setTouchEmulationEnabled', 
  {
    enabled: true,
    configuration: 'mobile',
  }, function(err){
     if(err){console.log("Emulation err:", err);}
  }
);
```

I develop a tool with electron. There are two webview components in the app. One shows the html page, another one shows the chrome devtools which debug the page. The current solution is:

const browserContents = browserView.getWebContents()
const devtoolsContents = devtoolsView.getWebContents()

browserContents.setDevToolsWebContents(devtoolsContents)
browserContents.openDevTools({
     detach: true
})

browserContents.debugger.attach(); // 
browserContents.debugger.sendCommand(    // not work
          'Emulation.setTouchEmulationEnabled', 
  {
    enabled: true,
    configuration: 'mobile',
  }, function(err){
     if(err){console.log("Emulation err:", err);}
  }
);
The code above can solve the problem, but the browserContents.debugger.sendCommand does not work. So, I want to ask if there are any good ideas about how to solve this question?

**Comment by lishoulong** [comment]
I also encounter the same question, looking for @MarshallOfSound your help.

**Comment by zcbenz** [comment]
The only work around I can think of, is to execute the command in Devtools WebContents with `executeJavaScript` directly:

```js
devtoolsContents.executeJavaScript('...')
```

But you have to figure out which devtools JavaScript API to use.

I don't know much about the multiple debugger support of devtools, so I'm not sure if it can solve this problem.

**Comment by jacklovepdf** [comment]
thanks a lot , i have solve the problem.

**Comment by xiaoyang3874** [comment]
> thanks a lot , i have solve the problem.
 I have the same question, how did you solved it?

**Comment by jacklovepdf** [comment]
just create a websocket server between  chrome devtools and debugger server, then you can send any cmd you want;

**Comment by cyqresig** [comment]
@jacklovepdf 
can u describe how to solve the problem much more detailed？

**Comment by xyytech** [comment]
thanks @jacklovepdf, I have solve the problem with electron3
1.on the main process, add this code: `app.commandLine.appendSwitch('remote-debugging-port', '8400')`;
2.`<webview id="browserView" src="about:blank" webPreferences="scrollBounce=1,enableBlinkFeatures=OverlayScrollbars" plugins></webview>`
3.when the browserWebview ready:
```javascript
browserView.addEventListener('dom-ready', () => {
  if (isBrowserWebviewReady) return
  isBrowserWebviewReady = true
  fetch('http://127.0.0.1:8400/json')
    .then(res => res.json())
    .then(res => {
      const target = res.find(child => child.type === 'webview' && child.url === 'about:blank')
      if (target) {
        devtoolsView.setAttribute('src', `http://127.0.0.1:8400${target.devtoolsFrontendUrl}`)
        const ws = new WebSocket(target.webSocketDebuggerUrl)
        ws.onopen = () => {
          ws.send(JSON.stringify({
            id: 1,
            method: 'Emulation.setEmitTouchEventsForMouse',
            params: {
              enabled: true,
              configuration: 'mobile'
            }
          }))
        }
        devtoolsView.addEventListener('dom-ready', () => {
          browserView.setAttribute('src', 'http://xxxx.com')
        })
      }
    })
})
```

**Comment by Nashorn** [comment]
Hi,
there is a problem with contents.setDevToolsWebContents(devToolsWebContents) and the example screenshot above. The dev tools pane is not resizable like the default devtools allows. Any way to fix this?

**Comment by zhaomenghuan** [comment]
thanks @jacklovepdf and @xyytech，I have solve the some problem。

main.js:
```
app.commandLine.appendSwitch('remote-debugging-port', '8090');
```

renderer.js:
```
const baseUrl = 'http://127.0.0.1:8090';
const simulatorView = document.getElementById('simulator');
const devtoolsView = document.getElementById('devtools');
simulatorView.addEventListener('dom-ready', () => {
  fetch(`${baseUrl}/json`)
    .then(res => res.json())
    .then(res => {
      const target = res.find(child => child.type === 'webview' && child.url === simulatorView.src);
      if (target) {
      devtoolsView.setAttribute('src', `${baseUrl}${target.devtoolsFrontendUrl}`);
      const simulatorContents = simulatorView.getWebContents();
      const devtoolsContents = devtoolsView.getWebContents();
      simulatorContents.setDevToolsWebContents(devtoolsContents);
      simulatorContents.debugger.attach();
      simulatorContents.openDevTools({
        mode: 'detach'
      });
    }
  });
});
```

**Comment by Nashorn** [comment]
> thanks @jacklovepdf and @xyytech，I have solve the some problem。
> 
> main.js:
> 
> ```
> app.commandLine.appendSwitch('remote-debugging-port', '8090');
> ```
> 
> renderer.js:
> 
> ```
> const baseUrl = 'http://127.0.0.1:8090';
> const simulatorView = document.getElementById('simulator');
> const devtoolsView = document.getElementById('devtools');
> simulatorView.addEventListener('dom-ready', () => {
>   fetch(`${baseUrl}/json`)
>     .then(res => res.json())
>     .then(res => {
>       const target = res.find(child => child.type === 'webview' && child.url === simulatorView.src);
>       if (target) {
>       devtoolsView.setAttribute('src', `${baseUrl}${target.devtoolsFrontendUrl}`);
>       const simulatorContents = simulatorView.getWebContents();
>       const devtoolsContents = devtoolsView.getWebContents();
>       simulatorContents.setDevToolsWebContents(devtoolsContents);
>       simulatorContents.debugger.attach();
>       simulatorContents.openDevTools({
>         mode: 'detach'
>       });
>     }
>   });
> });
> ```

Has anyone tried this solution successfully?

**Pull request merged**